docs/pages/07_service.adoc

```diff
@@ -0,0 +1,1092 @@
+
+*Permission graph*
+
+// https://en.wikipedia.org/wiki/Graph-based_access_control
+
+JUDO's security mechanism is based on a logical permission graph, where the entry point for external users to the application is the actor. Calling the functions exported by the actor, the user can access transfer objects. Once the user has access to a transfer object, they can call the services exported by the transfer object (and can access further transfer objects).
+
+The primary benefit of this mechanism is that it provides a direct mapping between the user's permissions and the transfer objects and functions they can access. This makes it easier to manage permissions and control access to the application.
+
+The example below shows how a transfer object can export services. Users who received the PersonTransfer object can call the delete function.  This function can only be called by the users who have access to the PersonTransfer object.
+
+*Syntax:*
+
+----
+transfer <name> (<entity> <alias>) [exports <service>, ...] [{
+    [member] ...
+}]
+----
+
+*Example:*
+
+----
+transfer PersonTransfer(Person person) exports PersonService {
+    field String email = salesPerson.email ;
+}
+
+service PersonService(Person person) {
+    function void delete() {
+        // implement this body in Java
+    }
+}
+----
+
+The complete authorization mechanism can be defined by carefully provisioning services to actors and transfer objects. Always be aware of that if a transfer object has already been received by an external user, then they are authorized to call the functions provided by the transfer object.
+
+=== Guard
+
+A guard is a boolean expression that must evaluate to true in order to continue the execution.
+
+Guards can be defined for
+
+* functions,
+* services and
+* actors.
+
+To add a guard to functions use `guard` as follows:
+
+*Example:*
+
+----
+service PersonService(Person person) {
+    function void delete() guard person.leads!size() == 0;
+}
+----
+
+The example above defines a guard, and the delete function is not executed if that person has a lead.
+
+// TODO: what is the error message?
+
+To add a guard to services use `guard` as follows:
+
+*Example:*
+
+----
+service PersonService(Person person) guard person.leads!size() == 0 {
+    function void delete() ;
+    function Lead[] addLead(Lead lead) ;
+}
+----
+
+The example above defines a guard for service, and the functions are not executed and run-time error is thrown if that person has a lead.
+
+// TODO: add error definition to guards if error is intruduced
+
+To add a guard to actors use `guard` as follows:
+
+*Example:*
+
+----
+actor SalesPersonActor(SalesPerson salesPerson)
+    realm "COMPANY"
+    identity salesPerson.email
+    guard salesPerson.leads!size() > 0;
+----
+
+The example above defines a guard for actor, and the salesperson can only invoke the functions if she has at least one lead.
```

Comment:
I wouldn't support this, because I can already write it in the service and fuction level. The result of this concept will be more mistake and errors in the applications. In this case very easy to write contradictory guards, and the finding the mistake will be hard.

Reply:
ACK